components/PlaybackTask.brs

```diff
@@ -0,0 +1,253 @@
+  m.adFacade = m.top.adFacade
+  m.skipAds = false
+  m.lastPosition = 0
+  m.videoPlayer = invalid
+  m.raf = invalid
+  m.truexAd = invalid
+  m.currentAdPod = invalid
+end sub
+
+sub setupVideo()
+  ? "TRUE[X] >>> setupVideo()"
+
+  videoPlayer = m.top.video
+  videoContent = createObject("roSGNode", "ContentNode")
+
+  videoContent.url = "http://ctv.truex.com.s3.amazonaws.com/assets/reference-app-stream-no-ads-720p.mp4"
+  videoContent.length = 22 * 60
+
+  videoContent.title = "video title"
+  videoContent.streamFormat = "mp4"
+  videoContent.playStart = 0
+
+  videoPlayer.content = videoContent
+  videoPlayer.SetFocus(true)
+  videoPlayer.visible = true
+  videoPlayer.observeField("position", m.port)
+  videoPlayer.EnableCookies()
+
+  m.videoPlayer = videoPlayer
+end sub
+
+sub setupRaf()
+  ? "TRUE[X] >>> setupRaf()"
+  raf = Roku_Ads()
+  raf.enableAdMeasurements(true)
+  raf.setContentGenre("Entertainment")
+  raf.setContentId("TrueXSample")
+
+  raf.SetDebugOutput(false) 'debugging
+  raf.SetAdPrefs(false)
+
+  adUrl = m.top.adUrl
+  if adUrl = invalid OR adUrl = ""
+    ' adUrl = "pkg:/res/adpods/vast-regular.xml"  ' Regular Ads
+    adUrl = "pkg:/res/adpods/vmap-truex.xml" ' Preroll + Midroll Truex experience
+    ' adUrl = "pkg:/res/adpod/truex-pod-preroll.xml"
+    ' adUrl = "pkg:/res/adpod/truex-pod-skip.xml" ' Forced standalone skip pod.  Can be inserted in a full ad payload
+  end if 
+  raf.setAdUrl(adUrl)
+
+  m.raf = raf
+end sub
+
+sub setupEvents()
+  m.top.observeField("exitPlayback", m.port)
+end sub
+
+sub initPlayback()
+  ? "TRUE[X] >>> initPlayback()"
+
+  m.currentAdPod = getPreroll()
+  startPlayback() ' Will handle playing a preroll if it exists per above
+
+  while(true)
+    msg = Wait(0, m.port)
+    msgType = type(msg)
+
+    ' ? "event type:" msgType
+    if msgType = "roSGNodeEvent"
+      field = msg.getField()
+      ' ? "roSGNodeEvent msg.getField()" field
+      if field = "position" then 
+        onPositionChanged(msg)
+      else if field = "event" then
+        onTrueXEvent(msg)
+      else if field = "exitPlayback"
+        exitPlayback()
+      end if
+      ' TODO: Video complete event
+    end if
+  end while
+end sub
+
+sub onPositionChanged(event)
+  m.lastPosition = event.getData()
+
+  ads = m.raf.getAds(event)
+  handleAds(ads)
+end sub
+
+sub onTruexEvent(event)
+  data = event.getData()
+  eventType = data.type
+
+  ? "TRUE[X] >>> PlaybackTask::onTrueXEvent(): " eventType
+
+  types = {
+    "ADSTARTED": "adStarted",
+    "ADCOMPLETED": "adCompleted"
+    "ADPOSITION": "adPosition",
+    "ADERROR": "adError",
+    "ADFREEPOD": "adFreePod"
+    "NOADSAVAILABLE": "noAdsAvailable",
+    "OPTIN": "optIn",
+    "OPTOUT": "optOut",
+    "USERCANCEL": "userCancel",
+    "USERCANCELSTREAM": "userCancelStream",
+    "SKIPCARDSHOWN": "skipCardShown"
+  }
+
+  startPlaybackEvents = [types.ADCOMPLETED, types.NOADSAVAILABLE, types.ADERROR]
+
+  if eventType = types.ADFREEPOD
+    m.skipAds = true
```

Comment:
Good point.  I was mentally using the Sponsored Stream case as my primary use case.  But this technically would break for non-sponsored stream cases.  Will take a closer look at this, have some simple ideas.